File: Scripts/StockViewer/backend/app/schemas/portfolio.py
from datetime import datetime, date
from pydantic import BaseModel
from typing import Optional


class HoldingCreate(BaseModel):
    ticker: str
    market: str = "US"
    shares: float
    avg_cost: float
    currency: str = "USD"


class HoldingRead(BaseModel):
    id: int
    ticker: str
    market: str
    shares: float
    avg_cost: float
    currency: str

    model_config = {"from_attributes": True}


class PortfolioCreate(BaseModel):
    name: str


class PortfolioRead(BaseModel):
    id: int
    name: str
    created_at: datetime
    holdings: list[HoldingRead] = []

    model_config = {"from_attributes": True}


class PortfolioSummary(BaseModel):
    id: int
    name: str
    holding_count: int

    model_config = {"from_attributes": True}


class PnLPoint(BaseModel):
    date: date
    value: float


class PnLResponse(BaseModel):
    series: list[PnLPoint]
    total_invested: float
    current_value: float
    pnl_pct: Optional[float]


class OptimizeResult(BaseModel):
    weights: dict[str, float]
    expected_return: float
    volatility: float
    sharpe: float

File: Scripts/StockViewer/frontend/src/types/portfolio.ts
export interface Holding {
  id: number;
  ticker: string;
  market: string;
  shares: number;
  avg_cost: number;
  currency: string;
}

export interface Portfolio {
  id: number;
  name: string;
  created_at: string;
  holdings: Holding[];
}

export interface PortfolioSummary {
  id: number;
  name: string;
  holding_count: number;
}

export interface PnLPoint {
  date: string;
  value: number;
}

export interface PnLResponse {
  series: PnLPoint[];
  total_invested: number;
  current_value: number;
  pnl_pct: number | null;
}

export interface OptimizeResult {
  weights: Record<string, number>;
  expected_return: number;
  volatility: number;
  sharpe: number;
}
